networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BeaconBlocksByRootMessageHandler.java

```diff
@@ -55,11 +66,40 @@ public void onIncomingMessage(
                         .getStore()
                         .retrieveSignedBlock(blockRoot.get())
                         .thenCompose(
-                            block -> block.map(callback::respond).orElse(SafeFuture.COMPLETE)));
+                            block -> {
+                              final Optional<RpcException> validationResult =
+                                  block.flatMap(b -> validateResponse(protocolId, b));
+                              if (validationResult.isPresent()) {
+                                callback.completeWithErrorResponse(validationResult.get());
+                                return SafeFuture.failedFuture(validationResult.get());
+                              }
+                              return block.map(callback::respond).orElse(SafeFuture.COMPLETE);
+                            }));
       }
       future.finish(callback::completeSuccessfully, callback::completeWithUnexpectedError);
     } else {
       callback.completeSuccessfully();
     }
   }
+
+  @VisibleForTesting
+  Optional<RpcException> validateResponse(
+      final String protocolId, final SignedBeaconBlock response) {
+    final int version = BeaconChainMethodIds.extractBeaconBlocksByRootVersion(protocolId);
+    final Optional<UInt64> altairActivationEpoch =
+        spec.getForkSchedule().getMilestoneActivationEpoch(SpecMilestone.ALTAIR);
+    if (altairActivationEpoch.isEmpty()) {
+      return Optional.empty();
+    }
+
+    final UInt64 altairActivationSlot = spec.computeStartSlotAtEpoch(altairActivationEpoch.get());
+    final UInt64 responseSlot = response.getSlot();
```

Comment:
Feels like we have some duplication here in determining the first Altair slot for the two RPC methods.  I'm not exactly sure where it would live but a utility method to as "is this slot after the specified milestone activated" seems useful.  Probably `ForkSchedule.isMilestoneActiveAtSlot(milestone, slot)`?

Reply:
I couldn't come up with an unambiguous method to add to `ForkSchedule`, but I did simplify a bit.